README.md
# keygraph

The goal of this project is to implement the KeyGraph algorithm for chance discovery (Ohsawa et al., 1998; Ohsawa, 2006). It builds on the project https://github.com/ShinsakuSegawa/keygraph.

## Installation

Clone the repository, then run `setup.py` to install the required NLTK resources.

To render a KeyGraph, you also need to install a copy of  `Graphviz` (http://www.graphviz.org).

## Usage

Suppose the document you want to analyze is in the file `d1.txt` in the `txt_files` folder. To create a KeyGraph from the text in this document, run:

```bash
python3 keygraph.py d1
```

This creates input for `Graphviz` in the `dot` folder. To render the KeyGraph, run:

```bash
dot -Tpdf dot/d1.dot -o graphs/d1.pdf
```

When generating the KeyGraph, stopwords in the `noise\stopwords.txt` file are used to remove noise words. To add stopwords, add one stopword or symbol per line.

## References

Ohsawa, Y., Benson, N. E., & Yachida, M. (1998). KeyGraph: Automatic indexing by co-occurrence graph based on building construction metaphor. International Forum on Research and Technology Advances in Digital Libraries (ADL), 12-18). IEEE.

Ohsawa, Y. (2006). Chance discovery: The current states of art. In: Ohsawa, Y., & Tsumoko, E. (eds.), Chance Discoveries in Real World Decision Making, 3-20. Springer.

txt_files/news.txt
New evidence emerged in the impeachment trial about what President Donald J. Trump did from roughly 1 to 6 p.m. the day of the Capitol attack. But many questions remain unanswered.

Published Feb. 13, 2021Updated Feb. 14, 2021, 12:09 a.m. ET
President Donald J. Trump at a rally near the White House on Jan. 6, the day of the Capitol siege.
President Donald J. Trump at a rally near the White House on Jan. 6, the day of the Capitol siege.Pete Marovich for The New York Times
The impeachment trial of former President Donald J. Trump largely focused on his actions leading up to the violent attack on the Capitol on Jan. 6. But there was a crucial period that day of nearly five hours — between the end of Mr. Trump’s speech at the Ellipse urging his supporters to march to the Capitol and a final tweet telling his followers to remember the day forever — that remains critical to his state of mind.

Evidence emerged during the trial about what Mr. Trump was doing during those hours, including new details about two phone calls with lawmakers that prosecutors said clearly alerted the president to the mayhem on Capitol Hill. Prosecutors said the new information was clear proof of Mr. Trump’s intent to incite the mob and of his dereliction to stop the violence, even when he knew that the life of Vice President Mike Pence was in danger.

Senator Mitch McConnell of Kentucky, the Republican leader who on Saturday voted to acquit Mr. Trump but offered a sweeping endorsement of the prosecutors’ case, backed them up: “There’s no question — none — that President Trump is practically and morally responsible for provoking the events of the day. No question about it.”

Still, many crucial questions remain unanswered about the president’s actions and mood from roughly 1 to 6 p.m. Jan. 6. Here is what is known so far:

Mr. Trump concluded his incendiary speech on the Ellipse at 1:11 p.m. He had repeatedly told the crowd that the election was stolen from him and urged his supporters to march to the Capitol in a last-ditch effort to stop President-elect Joseph R. Biden Jr.’s victory from being certified. Mr. Trump said twice that he would go with them. And days before the march, he had told advisers that he wanted to join his supporters, but aides told him that people in the crowd were armed and that the Secret Service would not be able to protect him.

Six minutes later, Mr. Trump’s motorcade began heading back to the White House. He arrived there at 1:19 p.m. as the crowd was making its way up Pennsylvania Avenue and beginning to swarm around the Capitol. Television news footage showed the mob as it moved closer to the doors.

At some point, Mr. Trump went to the Oval Office and watched news coverage of a situation that was growing increasingly tense.


How Trump’s Calls to G.O.P. Lawmakers Fit in the Timeline of the Capitol Riot

Here is a moment-by-moment timeline of events on Jan. 6, when a mob assaulted the U.S. Capitol.

At 1:34 p.m., Mayor Muriel Bowser of Washington made a formal request for assistance in a phone call with the Army secretary, Ryan D. McCarthy. At 1:49 p.m., as the Capitol Police asked Pentagon officials for help from the National Guard, Mr. Trump tweeted a video of his incendiary rally speech.

It was around this time that some of Mr. Trump’s allies publicly called on him to do something. Chris Christie, the former governor of New Jersey, told ABC News that Mr. Trump needed to say something to stop the rioting.

At 2:12 p.m., the same moment that the mob breached the building itself, Mr. Pence — who had defied the president by saying he planned to certify Mr. Biden’s victory — was rushed off the Senate floor. A minute later, the Senate session was recessed. Two minutes after that, at 2:15 p.m., groups of rioters began to chant, “Hang Mike Pence!”

Nine minutes later, at 2:24 p.m., Mr. Trump tweeted a broadside at Mr. Pence for moving ahead to certify Mr. Biden’s win: “Mike Pence didn’t have the courage to do what should have been done to protect our Country and our Constitution, giving States a chance to certify a corrected set of facts, not the fraudulent or inaccurate ones which they were asked to previously certify. USA demands the truth!”

At 2:26 p.m., after Mr. Pence had been whisked away, a call was placed from the White House to Senator Mike Lee, Republican of Utah, according to call logs that the senator provided during the impeachment proceedings.

The president had made the call, but he was actually looking for Senator Tommy Tuberville, Republican of Alabama. Mr. Lee gave the phone to Mr. Tuberville, who has told reporters that he informed Mr. Trump that Mr. Pence had just been escorted out as the mob got closer to the Senate chamber.

“I said, ‘Mr. President, they just took the vice president out, I’ve got to go,’” Mr. Tuberville recounted to Politico.

This was a significant new piece of information. House prosecutors used it to argue that Mr. Trump was clearly aware that the vice president was in danger and that he had a callous disregard for Mr. Pence’s safety. On Friday, Mr. Trump’s defense team had insisted that Mr. Trump was not aware of any peril facing Mr. Pence.

Back at the White House, advisers were trying to get Mr. Trump to do something, but he rebuffed calls to intercede, including those from people wanting to see the National Guard deployed. The president, several advisers said, was expressing pleasure that the vote to certify Mr. Biden’s win had been delayed and that people were fighting for him.

“According to public reports, he watched television happily — happily — as the chaos unfolded,” Mr. McConnell said on Saturday. “He kept pressing his scheme to overturn the election. Even after it was clear to any reasonable observer that Vice President Pence was in serious danger, even as the mob carrying Trump banners was beating cops and breaching perimeters, the president sent a further tweet attacking his own vice president.”


test.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sun Feb 14 17:05:16 2021

@author: mrw
"""

import nltk
from document import Document

def test_lemmatize():
    lemmatizer = nltk.stem.WordNetLemmatizer()
    for w in ["last", "las", "us", "Q1-1", "decreased"]:
        print(w, lemmatizer.lemmatize(w))
        
def test_strip():
    for s in [" a", "b ", " c ", "d\n", "e"]:
        print("<{}>".format(s), "<{}>".format(s.strip()))
        
def test_pos_tag():
    s = "the wealthy customers decreased"
    s = "But that may reduce us the power of this company."
    tags = nltk.pos_tag(nltk.word_tokenize(s))
    print(tags)
    print([(w, wordnet_pos(t)) for w, t in tags])
    
def test_lemmatize_pos():
    s = "the wealthy customers decreased"
    s = "But that may reduce us the power of this company."
    lemmatizer = nltk.stem.WordNetLemmatizer()
    tags = nltk.pos_tag(nltk.word_tokenize(s))
    print([lemmatizer.lemmatize(w, wordnet_pos(t)) for w, t in tags])
   
def test_tokens():
    doc = Document(file_name = "./txt_files/d2.txt")
    tokens = doc.create_tokens()
    print(tokens)
    
def test_sentences():
    doc = Document(file_name = "./txt_files/d2.txt")
    sentences = doc.create_sentences()
    print(sentences)
    
def test_tokens_from():
    doc = Document("the brown foxes jumped over the fence")
    tokens = doc.create_tokens_from(doc.content, lemmatized=False, strip_stopwords=False)
    print(tokens)
    tokens = doc.create_tokens_from(doc.content, lemmatized=False)
    print(tokens)
    tokens = doc.create_tokens_from(doc.content)
    print(tokens)
    
# Lookup WordNet POS
# https://www.machinelearningplus.com/nlp/lemmatization-examples-python/
def wordnet_pos(tag):
    tags = {"J": nltk.corpus.wordnet.ADJ,
            "N": nltk.corpus.wordnet.NOUN,
            "V": nltk.corpus.wordnet.VERB,
            "R": nltk.corpus.wordnet.ADV}
    # tag example: 'VBD' for verb
    return tags.get(tag[0], nltk.corpus.wordnet.NOUN)
        
#-----------Main----------------
if __name__ == "__main__":
    test_lemmatize()
    test_strip()
    test_pos_tag()
    test_lemmatize_pos()
    test_tokens()
    test_sentences()
    test_tokens_from()


pygraphviz_ja_test.py
# coding=UTF-8
from pygraphviz import *
import codecs

f = codecs.open("./dot/base.dot","r","utf-8")

base = ""
for row in f:
    base += row

print base
'''
 
dot = """
graph graph1
{
              猫 -- ドラえもん -- 狸;
                札幌 -- 富良野;
                }
"""
 
'''
 
graph = AGraph(string=base)
graph.layout(prog='dot')
graph.draw('./graphs/pgv1.png')


key.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys
# import MeCab
import codecs
import pprint
import time
import nltk
 
M = 20
K = 12

# sys.stdout = codecs.getwriter('utf_8')(sys.stdout)
# sys.stdin = codecs.getreader('utf_8')(sys.stdin)

# Pretty-print a Python object
def pp(obj):
    pp = pprint.PrettyPrinter(indent=4, width=160)
    s = pp.pformat(obj)
    return s
  
# Read file name from the console
def get_file_name():
    if (len(sys.argv) != 2):
        print("Usage: #python %s file-name" % sys.argv[0])
        sys.exit()
    print(pp(sys.argv))
    return sys.argv[1]

# Read document from file
def read_from_file(file_name):
    f = codecs.open('./txt_files/' + fname + '.txt', 'r', 'utf-8')
    doc = f.read()
    f.close()
    return doc

# Delete noise
def delNoise(text):
    return text 
 
# Divide into sentences
def creSentence(text):
    return [s.lower() for s in nltk.tokenize.sent_tokenize(text)]

# Divide into sentences and tokenize each sentence
def create_sentences(text):
    return [create_tokens(s) for s in creSentence(text)]

# Divide into tokens
def create_tokens(text):
    return [t.lower() for t in nltk.tokenize.word_tokenize(text)]

# Read user-defined stopwords
def read_user_defined_stopwords():
    stopwords = []
    for line in codecs.open('./noise/stopwords.txt', 'r', 'utf-8'):
        stopwords.append(line.strip())
    return stopwords

user_defined_stopwords = read_user_defined_stopwords()

# Strip stopwords and special symbols from list of words
def strip_stopwords_and_symbols(tokens):
    stopwords = nltk.corpus.stopwords.words('english')
    symbols = ["'", '"', '“', '”', '`', '’', '.', ',', '-', '!', '?', ':', ';', '(', ')', '[', ']', '&', '0', '%', '...', '--']
    return [w for w in tokens 
            if w not in stopwords + user_defined_stopwords + symbols and len(w) > 1]

# Lemmatize words
def lemmatize(tokens):
    lemmatizer = nltk.stem.WordNetLemmatizer()
    return [lemmatizer.lemmatize(w, wordnet_pos(t)) for w, t in nltk.pos_tag(tokens)]

# Lookup WordNet POS
# https://www.machinelearningplus.com/nlp/lemmatization-examples-python/
def wordnet_pos(tag):
    tags = {"J": nltk.corpus.wordnet.ADJ,
            "N": nltk.corpus.wordnet.NOUN,
            "V": nltk.corpus.wordnet.VERB,
            "R": nltk.corpus.wordnet.ADV}
    # tag example: 'VBD' for verb
    return tags.get(tag[0], nltk.corpus.wordnet.NOUN)
    
# Count word frequencies
def freqcount(tokens):
    result = {}
    for t in tokens:
        if t in result:
            result[t] += 1
        else:
            result[t] = 1
    return result	
 
#	Calculate word frequency in sentences
def calwfs(words, sents):
    wfs = {} 
    for w in words:
        for s_idx, s in enumerate(sents):
            if w not in wfs:
                wfs[w] = {}
            wfs[w][s_idx] = s.count(w)
    return wfs	
 
#   Calculate co-occurrence degree of high-frequency words
def calCo(hf, sents):
    co = {} 
    for hf1 in hf:
        co[hf1] = {}
        for hf2 in hf[hf.index(hf1)+1:]:
            co[hf1][hf2] = 0 
            for s in sents:
                # Why sum products, not min, as in Ohsawa (1998)?
                # co[hf1][hf2] += s.count(hf1) * s.count(hf2)
                co[hf1][hf2] += min(s.count(hf1), s.count(hf2))
    co_list = [] 
    for x in co.keys():
        for y in co[x].keys():
            co_list.append([x, y, co[x][y]])    
    co_list.sort(key=lambda a: a[2])
    return co_list 
 
# Compute key (terms that tie and hold clusters together) 
def key(words, wfs, base, sents):
    # key is a dictionary of the form　key = {w: key value}	
    key = {}   
    Fg = fg(words, wfs, base, sents)
    for w in words:
        product = 1.0
        for g in base:
            product *= 1 - fwg(w, wfs, g, sents)*(1.0)/Fg[g]
#        print("product", product)
        key[w] = 1.0 - product 
        print("key[{}]".format(w), 1.0 - product, w)        
    return key		

# Calculate f(w,g)
# Based(w, g) = how many times w appeared in D, based on the basic
# concept represented by term g
def fwg(w, wfs, g, sents):	
    gws = 0
    fwg = 0
#    print("w", w, "g", g)
    for s, sentence in enumerate(sents):
#        print("sentence", sentence)
        # Calculate |g-w|_s
        # Count of cluster term g in s: g_s = wfs[g][s]
        # Word in s: w_s = wfs[w][s]
        if w == g: # w ∈ g
#            print("|g-w|", g, wfs[g][s], w, wfs[w][s])
            gws = wfs[g][s] - wfs[w][s]
        else: # w not ∈ g
#            print("|g|", g, wfs[g][s])
            gws = wfs[g][s]
        fwg += wfs[w][s] * gws
    return fwg

# Calculate F(g)
# Neighbors(g) = count of terms in sentences including terms in cluster g
# g_s = count of cluster g in sentence s
# w_s = count of word w in sentence s (ie wfs[w][s])
def fg(words, wfs, base, sents):
    fg = {}
    for g in base:
        fg[g] = 0
        for s, sentence in enumerate(sents):
            for w in sentence:
                if w == g: # w ∈ g
#                    print("|g-w|", g, wfs[g][s], w, wfs[w][s])
                    fg[g] += wfs[g][s] - wfs[w][s]
                else: # w not ∈ g
#                    print("|g|", g, wfs[g][s])
                    fg[g] += wfs[g][s] 
    return fg

def C(hk, base, sents):
    c = {}
    for k in hk:
        c[k] = {}
        for b in base:
            c[k][b] = 0
            for s in sents:
                c[k][b] += min(s.count(k), s.count(b))
    c_list = [] 
    for x in c.keys():
        for y in c[x].keys():
            c_list.append([x, y, c[x][y]])
    c_list.sort(key=lambda a: a[2])
    return c_list 
  
# Prune graph by removing edges that connect clusters
# That is, if the two ends of an edge are connected only by
# one edge, remove the edge to create two clusters
def prune(base, base_adj):
    new_base = []
    for [i, j] in base:
        if find_path(base_adj, i, j, [i, j], []):
            print("find path:", [i, j])
            new_base.append([i, j])
        else:
            print("no path found:", [i, j])
    return new_base
     
# Find a path between two nodes that does not include edge
# This won't necessarily return the best path, but is enough
# to test whether there is a path
# https://www.python.org/doc/essays/graphs/
def find_path(graph, start, end, edge, path=[]):
    path = path + [start]
    if start == end:
        return path
    if not start in graph:
        return None
    for node in [g[0] for g in graph[start] if g[1] == 'base']:
        if (node not in path) and not([path[-1], node] == edge):
            new_path = find_path(graph, node, end, edge, path)
            if new_path:
                return new_path
    return None
        
# Draw keygraph in dot format
def draw(base, G_C, fname):
    fout = codecs.open("./dot/" + fname + ".dot","w","utf-8")
    fout.write('graph keygraph {\n')
    fout.write('graph [size="10,10"]\n')

    g = []
    for i, j in base:
        g.append(i)
        g.append(j)
    for i in set(g):
        fout.write(quote(i) + ' [color="black"]\n')
    k = []
    for i, j in G_C:
        k.append(i)
    for i in set(k):
        fout.write(quote(i) + ' [color="red"]\n')
        
    for i, j in base:
        fout.write(quote(i) + '--' + quote(j) +'\n')
    for i, j in G_C:
        fout.write(quote(i) + '--' + quote(j) + ' [color="red", style="dotted"]\n')
    fout.write('}')
    fout.close()
    
# Add optional quotes around a name
def quote(name):
    if 1 in [c in name for c in ['-', '/', '.', '\'']]:
        return "\"{}\"".format(name)
    return name

# Create an adjacency list 
def adjacency_dic(base, G_C, fname):
    a_dic = {}
     
    for i,j in base:
        if i in a_dic:
            a_dic[i].append([j,'base'])
        else:
            a_dic[i] = [[j,'base']] 
        if j in a_dic:
            a_dic[j].append([i,'base'])
        else:
            a_dic[j] = [[i,'base']] 
    
    for i, j in G_C:
        if i in a_dic:
            a_dic[i].append([j,'key'])
        else:
            a_dic[i] = [[j,'key']] 
        if j in a_dic:
            a_dic[j].append([i,'key'])
        else:
            a_dic[j] = [[i,'key']] 

    fout = codecs.open("./adjacency_list/" + fname + ".txt","w","utf-8")
    fout.write(pp(a_dic))
    fout.close()
    
    return a_dic
        
#-----------Main----------------
if __name__ == "__main__":
    stime = time.time() 
    
    fname = get_file_name()
    doc = read_from_file(fname)
    
#	Delete noise
    nc_text = delNoise(doc)
             
#	Divide into sentences
    sents = create_sentences(nc_text)
    
    sents = [lemmatize(s) for s in sents]
    sents = [strip_stopwords_and_symbols(s) for s in sents]

#    sents = creSentence(nc_text)
#    sents = lemmatize_tokens_in_sentences(sents)
            
#	Divide into tokens
    tokens = create_tokens(nc_text)
    tokens = lemmatize(tokens)
    tokens = strip_stopwords_and_symbols(tokens)
    
#	Count word frequencies	
    freq_dict = freqcount(tokens)
    
#   Sort words by their frequency (in ascending order)
    words_freq = sorted(freq_dict.items(), key=lambda x: x[1])

#   Compute unique words
    words = [w for w, z in words_freq]
          
#	Calculate word frequency in sentences
    wfs = calwfs(words, sents)
            
#	Determine high frequency words
    hf = [w for w, f in words_freq[-M:]]
               
#   Calculate co-occurrence degree of high-frequency words
    co = calCo(hf, sents)
        
#   Compute the base of G (links between black nodes)
    base = [[i, j] for i, j, c in co[-M:]]  
    
#   Extract nodes in the base
    G_base = set([x for pair in base for x in pair])
    
#   Remove high frequency words from, leaving non-high frequency words
    words = [w for w in words if w not in G_base]
    
    print(pp(co))
    print(pp(G_base))
    print(pp(words))
   
#   Compute key (terms that tie and hold clusters together) 
    key = key(words, wfs, G_base, sents)

#   Sort terms in D by keys (produces list of terms ranked by their
#   association with the cluster)
    high_key = sorted(key.items(), key=lambda x: x[1])
    high_key = high_key[-K:]
    
    high_key = [k for k, f in high_key]
    
#	Calculate columns c(wi,wj)
    C = C(high_key, G_base, sents)
    C.sort(key=lambda x: x[2])
     
#   Compute the top links between key terms (red nodes) and clusters
    G_C = [[i, j] for i, j, c in C[-K:]]  
    
    # for i, j in base:
    #     print(i, j)
    # for x, y in G_C:
    #     print(x, y)

    base_adj = adjacency_dic(base, G_C, fname)
    
    pruned_based = prune(base, base_adj)
    
    draw(pruned_based, G_C, fname)

    etime = time.time()
    print("Execution time: %.4f seconds" % (etime - stime))



document.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import codecs
import nltk

class Document:
    def __init__(self, content = "", file_name = None):
        self.content = self.read_from_file(file_name) if file_name else content
        self.stopwords = self.read_stopwords()
        self.tokens = self.create_tokens()
        self.sentences = self.create_sentences()
    
    # Read content from file
    def read_from_file(self, file_name):
        f = codecs.open(file_name, 'r', 'utf-8')
        content = f.read()
        f.close()
        return content
    
    # Read user-defined stopwords
    def read_stopwords(self):
        stopwords = nltk.corpus.stopwords.words('english')
        for line in codecs.open('./noise/stopwords.txt', 'r', 'utf-8'):
            stopwords.append(line.strip())
        symbols = ["'", '"', '“', '”', '`', '’', '.', ',', '-', '!', '?', ':', ';', '(', ')', '[', ']', '&', '0', '%', '...', '--']
        return stopwords + symbols

    # Divide a string into tokens
    def create_tokens_from(self, s, lemmatized=True, strip_stopwords=True):
        tokens = [t.lower() for t in nltk.tokenize.word_tokenize(s)]
        tokens = self.lemmatize(tokens) if lemmatized else tokens
        return self.strip_stopwords(tokens) if strip_stopwords else tokens
    
    # Strip stopwords and symbols from tokens
    def strip_stopwords(self, tokens):
        return [t for t in tokens if not t in self.stopwords and len(t) > 1]
        
    # Divide into tokens
    def create_tokens(self):
        return self.create_tokens_from(self.content)

    # Divide into sentences
    def create_sentences(self):
        return [self.create_tokens_from(s) for s in nltk.tokenize.sent_tokenize(self.content)]

    # Lemmatize words
    def lemmatize(self, tokens):
        lemmatizer = nltk.stem.WordNetLemmatizer()
        return [lemmatizer.lemmatize(w, self.wordnet_pos(t)) 
                for w, t in nltk.pos_tag(tokens)]
    
    # Lookup WordNet POS
    # https://www.machinelearningplus.com/nlp/lemmatization-examples-python/
    def wordnet_pos(self, tag):
        tags = {"J": nltk.corpus.wordnet.ADJ,
                "N": nltk.corpus.wordnet.NOUN,
                "V": nltk.corpus.wordnet.VERB,
                "R": nltk.corpus.wordnet.ADV}
        # tag example: 'VBD' for verb
        return tags.get(tag[0], nltk.corpus.wordnet.NOUN)
   
    # Count word frequencies
    def freq_count(self):
        result = {}
        for t in self.tokens:
            if t in result:
                result[t] += 1
            else:
                result[t] = 1
        return result
    
     


KeyGraph.py
#! python/keygraph/KeyGraph.py
# -*- coding: utf-8 -*-
import sys
import MeCab
import codecs
import re, pprint
import itertools
import time
import unicodedata
import os 
 
sys.stdout = codecs.getwriter('utf_8')(sys.stdout)
sys.stdin = codecs.getreader('utf_8')(sys.stdin)

def pp(obj):
    pp = pprint.PrettyPrinter(indent=4, width=160)
    str = pp.pformat(obj)
    return re.sub(r"\\u([0-9a-f]{4})", lambda x: unichr(int("0x"+x.group(1),16)),str)

  
M = 30

print 'start'

def readFiles():
    f_list = os.listdir('/Users/SS/python/keygraph/txt_files/')
    files = []
    for f in f_list:
        files.append(f.replace(".txt", ""))

    for f in files:
        print f

    return files


#ノイズの削除
def delNoise(text):
#   ノイズファイルの読み込み
    noise_list=[]
    for line in codecs.open('./noise/noise.txt', 'r', 'utf-8'):
        noise_list.append(line[:-1].split("\n"))	

#   ノイズ削除
    for i in range(len(noise_list)-1):
        noise =  noise_list[i]
        text = text.replace(noise[0], u'')
    return text 
 
#センテンス（文）のlist作成
def creSentence(text):
    s_cha = u'。'
#	    cat = re.split(ur"", text)
    sents = []
    sents = text.strip(s_cha).split(s_cha)
    return sents

#名詞だけを取得してリストにいれる　リストを返す 数字抜き
def pyMecab(s):
    tagger = MeCab.Tagger('-Ochasen')
    node = tagger.parseToNode(s.encode('utf-8'))
    nouns = []
    while node:
#		    めかぶの名詞のidが38から60なんだよね　数字id=48抜き
        if node.posid >= 38 and node.posid <= 67 and not(node.posid == 49):
            if isValid(node.surface.decode('utf-8')) == True:
                nouns.append(node.surface.decode('utf-8'))
        node = node.next
    return nouns


def isValid(word):
    alldigit = re.compile(ur"^[0-9]+$")
    """wordが登録対象の単語のときTrueを返す"""
# 	    1文字の単語は登録しない
    if len(word) == 1:
        return False
# 	    数字だけの単語は登録しない
    if alldigit.search(word) != None:
        return False
# 	    仮名2文字の単語は登録しない
    if len(word) == 2 and unicodedata.name(word[0])[0:8] == "HIRAGANA" and unicodedata.name(word[1])[0:8] == "HIRAGANA":
        return False
#       仮名、漢字、数字、英字以外の文字を含む単語は登録しない
    for c in word:
        if not (unicodedata.name(c)[0:8] == "HIRAGANA" or
               unicodedata.name(c)[0:8] == "KATAKANA" or
               unicodedata.name(c)[0:3] == "CJK" or
               unicodedata.name(c)[0:5] == "DIGIT" 
#               英語を削除しています　コメントをとれば英語も入ります． 
              ''' unicodedata.name(c)[0:5] == "LATIN"'''):
            return False
    return True
#   popで熟語のやつを消すぜー	
def pop(dic_p, dic_q):

    list_p = dic_p.items()
    list_p.sort(key=lambda a: a[1])
    list_q = dic_q.items()
    list_q.sort(key=lambda a: a[1])
     
    for kq,vq in list_q:
        for kp,vp in list_p:
#			含むかどうか確認
            if(kq.count(kp)>0):
#				print kp, di_comb_p[kp], kq, di_idiom_q[kq]
                if(dic_p.get(kp) <= dic_q.get(kq)):
                    if(dic_p.has_key(kp)):
                        #print kp, kq
                        dic_p.pop(kp)
                else:
                    if(dic_q.has_key(kq)):
                        #print kq, kp
                        dic_q.pop(kq)
            if(kp.count(kq)>0):
#				print kp, di_comb_p[kp], kq, di_idiom_q[kq]
                if(dic_p.get(kp) <= dic_q.get(kq)):
                    if(dic_p.has_key(kp)):
                        #print kp, kq
                        dic_p.pop(kp)
                else:
                    if(dic_q.has_key(kq)):
                        #print kq, kp
                        dic_q.pop(kq)
                         
                         
    list_p = dic_p.items()
    list_q = dic_q.items()
    list_idiom = list_p
    list_idiom.extend(list_q)
    list_p.sort(key=lambda a: a[1])

    return list_p


#[語(word):出現回数]の辞書を作る
def freqcount(tokens):
    result={}
#   単語の出現でカウント(list内のカウント）
    for t in tokens:
        if result.has_key(t):
            result[t] += 1
        else:
            result[t] = 1
    return result	
 
#リスト['word']['sentence　インデックス']=出現回数　を返す関数
def calwfs(words, sents):
    wfs = {} 
    for w in words:
        for s in sents:
            if w in wfs:
                wfs[w][s] = s.count(w)
            else:
                wfs[w] = {}
                wfs[w][s] = s.count(w) 

#	for w,n in words:
#		for s in range(len(sents)):
#			print wfs[w][s]	

    return wfs	
 
#共起度coを計算する
def calCo(hf,sents):
    co = {} 
    for hf1 in hf:
        co[hf1] = {} #初期化
        for hf2  in hf[hf.index(hf1)+1:]:
            co[hf1][hf2] = 0 #初期化だよ 
            for s in sents:
                co[hf1][hf2] += s.count(hf1) * s.count(hf2)
                 
#	listにしています	
    co_list = [] 
    for x in co.keys():
        for y in co[x].keys():
            co_list.append([x,y,co[x][y]])
    
    co_list.sort(key=lambda a: a[2])

    return co_list 
 
 
#linkを張る
def link(base):
    base_set = flatten(base)
    base_set = set(base_set)
         
    print list(base_set)
    return list(base_set)
     
#リストの平坦化 
def flatten(x, isflat=lambda x:not isinstance(x, list)):
    if isflat(x):
        yield x
    else:
        for item in x:
            for i in flatten(item, isflat):
                yield i
                            
#key値の計算
def key(words, wfs, base, sents):
#	keyは辞書型　key = {w:key値}	
    key = {}
     
    Fg = fg(words, wfs, base, sents)

    for w in words:
        tmp = 1.0
        tmp_count = 0
        for g in base:
            tmp *= (1-(fwg(w,wfs,g,sents)*(1.0)/Fg[g])) 
        key[w] = 1.0-tmp 
        print 1.0-tmp, w 
         
    return key		

#f(w,g)計算
def fwg(w,wfs,g,sents):	
    gws = 0
    fwg = 0
    for s in sents:
#		|g-w|sの計算  gs = wfs[g][s] ws = ws[w][s]
        if g.find(w) >= 0:# w ∈ g
#			|g|sの計算
            gws = wfs[g][s] - wfs[w][s]
        else:# w not ∈ g
            gws = wfs[g][s]
        fwg += wfs[w][s]*gws
    return fwg

#F(g)の計算
def fg(words, wfs, base, sents):
    fg = {}
    for g in base:
        fg[g] = 0
        for s in sents:
            for w in words:
                if s.find(w) >= 0:
                    if g.find(w) >= 0:# w ∈ g
                        fg[g] += wfs[g][s] - wfs[w][s]
                    else:# w not ∈ g
                        fg[g] += wfs[g][s] 
    return fg


def C(hk, base, sents):
    c = {}
    for k in hk:
        c[k] = {}
        for b in base:
            c[k][b] = 0
            for s in sents:
                    c[k][b] += s.count(k) * s.count(b)
    
          
#	listにしています	
    c_list = [] 
    for x in c.keys():
        for y in c[x].keys():
            c_list.append([x,y,c[x][y]])
    
    c_list.sort(key=lambda a: a[2])

    return c_list 
  
def draw(base, G_C,fname):
    fout = codecs.open("./dot/" + fname + ".dot","w","utf-8")
    fout.write('graph keygraph {\n')
    fout.write('graph [size="10,10"]\n')
    for i,j in base:
       fout.write(i + '--' + j +'\n')
    for i,j in G_C:
       fout.write(i + '--' + j + '[style="dotted"]\n')
    fout.write('}')
     
    fout.close()

#隣接リストを作ったけど微妙だね． 
def adjacency_dic(base, G_C, fname):
    a_dic = {}
     
    for i,j in base:
        if a_dic.has_key(i):
            a_dic[i].append([j,'base'])
        else:
            a_dic[i] = [[j,'base']] 
        if a_dic.has_key(j):
            a_dic[j].append([i,'base'])
        else:
            a_dic[j] = [[i,'base']] 
    
    for i,j in G_C:
        if a_dic.has_key(i):
            a_dic[i].append([j,'key'])
        else:
            a_dic[i] = [[j,'key']] 
        if a_dic.has_key(j):
            a_dic[j].append([i,'key'])
        else:
            a_dic[j] = [[i,'key']] 


    fout = codecs.open("./adjacency_list/" + fname + ".txt","w","utf-8")
    fout.write(pp(a_dic.items()))
    fout.close()

         
         
        
#-----------Main----------------
if __name__ == "__main__":
    stime = time.time() 
#   コンソールからファイル名を読み込む 
    argvs = sys.argv
    argc = len(argvs)
    if(argc != 2):
        print 'Usage: #python %s filename' % argvs[0]
        quit()
         
    print pp(argvs)
    fname = argvs[1]
    
    
    f = codecs.open('./txt_files/' + argvs[1] + '.txt', 'r', 'utf-8')
     
    text = f.read()
    f.close()

#	ノイズの削除
    nc_text = delNoise(text)
     
#	センテンスリストの作成
    sents = creSentence(nc_text)

#	形態素解析
    tokens = pyMecab(nc_text) 
       
#	単語-出現度の辞書作成	
    freq_dict = freqcount(tokens)
    
    words_freq = sorted(freq_dict.items(), key=lambda x:x[1])
     
    words_freq = pop(dict(words_freq), dict(words_freq)) 
     
    words = [w for w,z in words_freq]
     
#	リスト[words][センテンス番号] = 出現回数のリストwfs
    wfs = calwfs(words, sents)

#	HighFreqを決める
    hf = [w for w,f in words_freq[-30:]]

#   HighFreqの共起度を計算する 返り値リスト[wi,wj,co]
    co = calCo(hf,sents)
     
    del words[-30:]
     
    base = [[i,j] for i,j,c in co[-30:]]  

        
    #baseに入っているノードを返す      
    G_base = link(base)   
     
    print G_base 
    
#	keyの計算  
    key = key(words, wfs, G_base, sents)
 
#	high_keyの計算
    high_key = sorted(key.items(), key=lambda x:x[1])
    high_key = high_key[-12:]
    
    high_key = [k for k,f in high_key]
     

#	c(wi,wj)の計算 [hk, base, スコア]が返り値
    C = C(high_key, G_base, sents)	
     
    C.sort(key=lambda x:x[2])
     
    G_C = [[i,j] for i,j,c in C[-12:]]  
     
    for i,j in base:
        print i,j
    
    for x,y in G_C:
        print x,y
    
    draw(base, G_C, fname)

    adjacency_dic(base, G_C, fname)
         
    etime = time.time()
 

print etime - stime	

